reporter/reporter.go

```diff
@@ -0,0 +1,259 @@
+package reporter
+
+import (
+	"log"
+	"strconv"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	metrics "github.com/rcrowley/go-metrics"
+	"github.com/wavefronthq/wavefront-sdk-go/application"
+	"github.com/wavefronthq/wavefront-sdk-go/histogram"
+	wf "github.com/wavefronthq/wavefront-sdk-go/senders"
+)
+
+// WavefrontMetricsReporter report go-metrics to wavefront
+type WavefrontMetricsReporter interface {
+	Start()
+	Stop()
+	Report()
+	ErrorsCount() int64
+}
+
+type reporter struct {
+	sender       wf.Sender
+	application  application.Tags
+	source       string
+	prefix       string
+	addSuffix    bool
+	interval     time.Duration
+	ticker       *time.Ticker
+	percentiles  []float64              // Percentiles to export from timers and histograms
+	durationUnit time.Duration          // Time conversion unit for durations
+	metrics      map[string]interface{} // for Wavefron specific metrics tyoes, like Histograms
+	errors       chan error
+	errorsCount  int64
+	errorDebug   bool
+	stop         chan bool
+	autoStart    bool
+	running      bool
+}
+
+// Option allow WavefrontReporter customization
+type Option func(*reporter)
+
+// LogErrors tag for metrics
+func LogErrors(debug bool) Option {
+	return func(args *reporter) {
+		args.errorDebug = debug
+	}
+}
+
+// Source tag for metrics
+func Source(source string) Option {
+	return func(args *reporter) {
+		args.source = source
+	}
+}
+
+// ReportingInterval change the default 1 minute reporting interval.
+func ReportingInterval(interval time.Duration) Option {
+	return func(args *reporter) {
+		args.interval = interval
+	}
+}
+
+// DisableAutoStart prevent the Repotorte to start reporting when is cretaed.
+func DisableAutoStart() Option {
+	return func(args *reporter) {
+		args.autoStart = false
+	}
+}
+
+// Prefix for the metrics name
+func Prefix(prefix string) Option {
+	return func(args *reporter) {
+		args.prefix = strings.TrimSuffix(prefix, ".")
+	}
+}
+
+// AddSuffix add a metric suffix based on the metric type ('.count', '.value')
+func AddSuffix(addSuffix bool) Option {
+	return func(args *reporter) {
+		args.addSuffix = addSuffix
+	}
+}
+
+// New create a WavefrontMetricsReporter
+func New(sender wf.Sender, application application.Tags, setters ...Option) WavefrontMetricsReporter {
+	r := &reporter{
+		sender:       sender,
+		application:  application,
+		source:       "mySource",
```

Comment:
Let's default to the hostname: `os.Hostname()` and if that returns an error default to `go-metrics-wavefront`.